{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "# Databases and web interfaces\n",
    "\n",
    "    Requirements:\n",
    "    pip install pymysql\n",
    "    pip install sqlalchemy\n",
    "    pip install weblogo\n",
    "\n",
    "\n",
    "\n",
    "## Connecting to a remote database server\n",
    "\n",
    "    UCSC Genomes database\n",
    "    http://genome.ucsc.edu/cgi-bin/hgTables\n",
    "\n",
    "## Direct database access and object-oriented ORM method\n",
    "## Using a python library with an object-oriented API\n",
    "## Creating a web interface for a pipeline using Flask\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "\n",
    "# from ucsc.db import create_session\n",
    "# from ucsc.config import use_genome\n",
    "\n",
    "# session = create_session('hg19')\n",
    "# genome = use_genome('hg19')\n",
    "\n",
    "\n",
    "from pprint import pprint"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from sqlalchemy import *\n",
    "# from sqlalchemy.ext.automap import automap_base"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The typical usage of create_engine() is once per particular database URL, held globally for the lifetime of a single application process. A single Engine manages many individual DBAPI connections\n",
    "\n",
    "\n",
    "Itâ€™s usually required that a separate Engine be used for each child process. This is because the Engine maintains a reference to a connection pool that ultimately references DBAPI connections - these tend to not be portable across process boundaries\n",
    "\n",
    "Here I am disabling connection pooling by Engine in order to use it in a Jupyter workbook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "from sqlalchemy.pool import NullPool\n",
    "\n",
    "engine = create_engine('mysql+pymysql://genome@genome-mysql.cse.ucsc.edu/hg38', poolclass=NullPool)  # , echo=True"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The connection is an instance of Connection, which is a proxy object for an actual DBAPI connection. The DBAPI connection is retrieved from the connection pool at the point at which Connection is created."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Table: affyGnf1h\n",
      "Table: affyU133\n",
      "Table: affyU95\n",
      "Table: all_est\n",
      "Table: all_mrna\n",
      "Table: all_sts_primer\n",
      "Table: all_sts_seq\n",
      "Table: altLocations\n",
      "Table: altSeqLiftOverPsl\n",
      "Table: altSeqLiftOverPslP3\n",
      "Table: altSeqLiftOverPslP6\n",
      "Table: altSeqLiftOverPslP7\n",
      "Table: augustusGene\n",
      "Table: bigFiles\n",
      "Table: bioCycMapDesc\n",
      "Table: bioCycPathway\n",
      "Table: ccdsGene\n",
      "Table: ccdsInfo\n",
      "Table: ccdsKgMap\n",
      "Table: ccdsNotes\n",
      "Table: ceBlastTab\n",
      "Table: centromeres\n",
      "Table: cgapAlias\n",
      "Table: cgapBiocDesc\n",
      "Table: cgapBiocPathway\n",
      "Table: chainBosTau8\n",
      "Table: chainBosTau8Link\n",
      "Table: chainCanFam3\n",
      "Table: chainCanFam3Link\n",
      "Table: chainChlSab2\n",
      "Table: chainChlSab2Link\n",
      "Table: chainDanRer10\n",
      "Table: chainDanRer10Link\n",
      "Table: chainFelCat8\n",
      "Table: chainFelCat8Link\n",
      "Table: chainGalGal5\n",
      "Table: chainGalGal5Link\n",
      "Table: chainGalVar1\n",
      "Table: chainGalVar1Link\n",
      "Table: chainGorGor5\n",
      "Table: chainGorGor5Link\n",
      "Table: chainHg38Patch3\n",
      "Table: chainHg38Patch3Link\n",
      "Table: chainHg38Patch6\n",
      "Table: chainHg38Patch6Link\n",
      "Table: chainHg38Patch7\n",
      "Table: chainHg38Patch7Link\n",
      "Table: chainMacFas5\n",
      "Table: chainMacFas5Link\n",
      "Table: chainMicMur2\n",
      "Table: chainMicMur2Link\n",
      "Table: chainMm10\n",
      "Table: chainMm10Link\n",
      "Table: chainMonDom5\n",
      "Table: chainMonDom5Link\n",
      "Table: chainPanPan1\n",
      "Table: chainPanPan1Link\n",
      "Table: chainPanTro5\n",
      "Table: chainPanTro5Link\n",
      "Table: chainRheMac8\n",
      "Table: chainRheMac8Link\n",
      "Table: chainRn5\n",
      "Table: chainRn5Link\n",
      "Table: chainSelf\n",
      "Table: chainSelfLink\n",
      "Table: chainTarSyr2\n",
      "Table: chainTarSyr2Link\n",
      "Table: chainXenTro7\n",
      "Table: chainXenTro7Link\n",
      "Table: chromAlias\n",
      "Table: chromInfo\n",
      "Table: clinvarCnv\n",
      "Table: clinvarMain\n",
      "Table: cloneEndABC10\n",
      "Table: cloneEndABC11\n",
      "Table: cloneEndABC12\n",
      "Table: cloneEndABC13\n",
      "Table: cloneEndABC14\n",
      "Table: cloneEndABC16\n",
      "Table: cloneEndABC18\n",
      "Table: cloneEndABC20\n",
      "Table: cloneEndABC21\n",
      "Table: cloneEndABC22\n",
      "Table: cloneEndABC23\n",
      "Table: cloneEndABC24\n",
      "Table: cloneEndABC27\n",
      "Table: cloneEndABC7\n",
      "Table: cloneEndABC8\n",
      "Table: cloneEndABC9\n",
      "Table: cloneEndCH17\n",
      "Table: cloneEndCOR02\n",
      "Table: cloneEndCOR2A\n",
      "Table: cloneEndCTD\n",
      "Table: cloneEndRP11\n",
      "Table: cloneEndWI2\n",
      "Table: cloneEndbadEnds\n",
      "Table: cloneEndcoverageForward\n",
      "Table: cloneEndcoverageReverse\n",
      "Table: cloneEndmultipleMaps\n",
      "Table: cnvDevDelayCase\n",
      "Table: cnvDevDelayControl\n",
      "Table: coriellDelDup\n",
      "Table: cpgIslandExt\n",
      "Table: cpgIslandExtUnmasked\n",
      "Table: crisprRanges\n",
      "Table: crisprTargets\n",
      "Table: ctgPos2\n",
      "Table: cytoBand\n",
      "Table: cytoBandIdeo\n",
      "Table: dbNsfpGerpNr\n",
      "Table: dbNsfpGerpRs\n",
      "Table: dbNsfpInterPro\n",
      "Table: dbNsfpLrt\n",
      "Table: dbNsfpMutationAssessor\n",
      "Table: dbNsfpMutationTaster\n",
      "Table: dbNsfpPolyPhen2\n",
      "Table: dbNsfpSeqChange\n",
      "Table: dbNsfpSift\n",
      "Table: dbNsfpUniProt\n",
      "Table: dbNsfpVest\n",
      "Table: dgvMerged\n",
      "Table: dgvSupporting\n",
      "Table: dmBlastTab\n",
      "Table: drBlastTab\n",
      "Table: estOrientInfo\n",
      "Table: extFile\n",
      "Table: extHg38Patch3\n",
      "Table: extHg38Patch6\n",
      "Table: extHg38Patch7\n",
      "Table: extNcbiAltSequence\n",
      "Table: fishClones\n",
      "Table: foldUtr3\n",
      "Table: foldUtr5\n",
      "Table: gap\n",
      "Table: gbLoaded\n",
      "Table: gc5BaseBw\n",
      "Table: geneReviews\n",
      "Table: geneReviewsDetail\n",
      "Table: geneid\n",
      "Table: genomicSuperDups\n",
      "Table: genscan\n",
      "Table: gnfAtlas2\n",
      "Table: gnfAtlas2Distance\n",
      "Table: gnfU95Distance\n",
      "Table: gold\n",
      "Table: grcIncidentDb\n",
      "Table: grp\n",
      "Table: gtexDistance\n",
      "Table: gtexGene\n",
      "Table: gtexGeneModel\n",
      "Table: gwasCatalog\n",
      "Table: gwipsvizRiboseq\n",
      "Table: hg38ContigDiff\n",
      "Table: hg38Patch3Haplotypes\n",
      "Table: hg38Patch3Patches\n",
      "Table: hg38Patch6Haplotypes\n",
      "Table: hg38Patch6Patches\n",
      "Table: hg38Patch7Haplotypes\n",
      "Table: hg38Patch7Patches\n",
      "Table: hgFindSpec\n",
      "Table: hgIkmc\n",
      "Table: hgIkmcExtra\n",
      "Table: history\n",
      "Table: humanHprdP2P\n",
      "Table: humanVidalP2P\n",
      "Table: humanWankerP2P\n",
      "Table: intronEst\n",
      "Table: iscaBenign\n",
      "Table: iscaBenignGainCum\n",
      "Table: iscaBenignLossCum\n",
      "Table: iscaCuratedBenign\n",
      "Table: iscaCuratedPathogenic\n",
      "Table: iscaLikelyBenign\n",
      "Table: iscaLikelyPathogenic\n",
      "Table: iscaPathGainCum\n",
      "Table: iscaPathLossCum\n",
      "Table: iscaPathogenic\n",
      "Table: iscaUncertain\n",
      "Table: keggMapDesc\n",
      "Table: keggPathway\n",
      "Table: kg7ToKg8\n",
      "Table: kg8ToKg9\n",
      "Table: kg9ToKg10\n",
      "Table: kgAlias\n",
      "Table: kgColor\n",
      "Table: kgProtAlias\n",
      "Table: kgSpAlias\n",
      "Table: kgTargetAli\n",
      "Table: kgXref\n",
      "Table: kgXrefOld8\n",
      "Table: kgXrefOld9\n",
      "Table: knownAlt\n",
      "Table: knownAttrs\n",
      "Table: knownBlastTab\n",
      "Table: knownCanonical\n",
      "Table: knownCds\n",
      "Table: knownGene\n",
      "Table: knownGeneMrna\n",
      "Table: knownGeneOld8\n",
      "Table: knownGeneOld9\n",
      "Table: knownGenePep\n",
      "Table: knownIsoforms\n",
      "Table: knownToEnsembl\n",
      "Table: knownToGencodeV20\n",
      "Table: knownToGencodeV22\n",
      "Table: knownToGencodeV24\n",
      "Table: knownToGnf1h\n",
      "Table: knownToGnfAtlas2\n",
      "Table: knownToGtex\n",
      "Table: knownToHprd\n",
      "Table: knownToKeggEntrez\n",
      "Table: knownToLocusLink\n",
      "Table: knownToLynx\n",
      "Table: knownToMrna\n",
      "Table: knownToMrnaSingle\n",
      "Table: knownToNextProt\n",
      "Table: knownToPfam\n",
      "Table: knownToRefSeq\n",
      "Table: knownToSuper\n",
      "Table: knownToTag\n",
      "Table: knownToU133\n",
      "Table: knownToU95\n",
      "Table: knownToVisiGene\n",
      "Table: knownToWikipedia\n",
      "Table: lincRNAsCTAdipose\n",
      "Table: lincRNAsCTAdrenal\n",
      "Table: lincRNAsCTBrain\n",
      "Table: lincRNAsCTBrain_R\n",
      "Table: lincRNAsCTBreast\n",
      "Table: lincRNAsCTColon\n",
      "Table: lincRNAsCTForeskin_R\n",
      "Table: lincRNAsCTHeart\n",
      "Table: lincRNAsCTKidney\n",
      "Table: lincRNAsCTLiver\n",
      "Table: lincRNAsCTLung\n",
      "Table: lincRNAsCTLymphNode\n",
      "Table: lincRNAsCTOvary\n",
      "Table: lincRNAsCTPlacenta_R\n",
      "Table: lincRNAsCTProstate\n",
      "Table: lincRNAsCTSkeletalMuscle\n",
      "Table: lincRNAsCTTestes\n",
      "Table: lincRNAsCTTestes_R\n",
      "Table: lincRNAsCTThyroid\n",
      "Table: lincRNAsCTWhiteBloodCell\n",
      "Table: lincRNAsCThLF_r1\n",
      "Table: lincRNAsCThLF_r2\n",
      "Table: lincRNAsTranscripts\n",
      "Table: locusName\n",
      "Table: lrg\n",
      "Table: lrgCdna\n",
      "Table: lrgCds\n",
      "Table: lrgPep\n",
      "Table: lrgTranscriptAli\n",
      "Table: malacards\n",
      "Table: mgcFullMrna\n",
      "Table: mgcGenes\n",
      "Table: microsat\n",
      "Table: mmBlastTab\n",
      "Table: mrnaOrientInfo\n",
      "Table: multiz100way\n",
      "Table: multiz100wayDefault\n",
      "Table: multiz100wayFrames\n",
      "Table: multiz100waySummary\n",
      "Table: multiz20way\n",
      "Table: multiz20wayFrames\n",
      "Table: multiz20waySummary\n",
      "Table: multiz7way\n",
      "Table: multiz7wayFrames\n",
      "Table: multiz7waySummary\n",
      "Table: nestedRepeats\n",
      "Table: netBosTau8\n",
      "Table: netCanFam3\n",
      "Table: netChlSab2\n",
      "Table: netDanRer10\n",
      "Table: netFelCat8\n",
      "Table: netGalGal5\n",
      "Table: netGalVar1\n",
      "Table: netGorGor5\n",
      "Table: netHg38Patch3\n",
      "Table: netHg38Patch6\n",
      "Table: netHg38Patch7\n",
      "Table: netMacFas5\n",
      "Table: netMicMur2\n",
      "Table: netMm10\n",
      "Table: netMonDom5\n",
      "Table: netPanPan1\n",
      "Table: netPanTro5\n",
      "Table: netRheMac8\n",
      "Table: netRn5\n",
      "Table: netTarSyr2\n",
      "Table: netXenTro7\n",
      "Table: omim2gene\n",
      "Table: omimAv\n",
      "Table: omimAvSnp\n",
      "Table: omimGene2\n",
      "Table: omimGeneMap\n",
      "Table: omimGeneSymbol\n",
      "Table: omimLocation\n",
      "Table: omimPhenotype\n",
      "Table: oreganno\n",
      "Table: oregannoAttr\n",
      "Table: oregannoLink\n",
      "Table: orfeomeGenes\n",
      "Table: orfeomeMrna\n",
      "Table: pfamDesc\n",
      "Table: phastCons100way\n",
      "Table: phastCons20way\n",
      "Table: phastCons7way\n",
      "Table: phastConsElements100way\n",
      "Table: phastConsElements20way\n",
      "Table: phastConsElements7way\n",
      "Table: phyloP100way\n",
      "Table: phyloP20way\n",
      "Table: phyloP7way\n",
      "Table: refFlat\n",
      "Table: refGene\n",
      "Table: refSeqAli\n",
      "Table: rmsk\n",
      "Table: rmskAlignBaseline\n",
      "Table: rmskJoinedBaseline\n",
      "Table: rmskOutBaseline\n",
      "Table: rnBlastTab\n",
      "Table: scBlastTab\n",
      "Table: scaffolds\n",
      "Table: scopDesc\n",
      "Table: seq\n",
      "Table: seqHg38Patch3\n",
      "Table: seqHg38Patch6\n",
      "Table: seqHg38Patch7\n",
      "Table: seqNcbiAltSequence\n",
      "Table: sgpGene\n",
      "Table: sibGene\n",
      "Table: sibTxGraph\n",
      "Table: simpleRepeat\n",
      "Table: snp141\n",
      "Table: snp141CodingDbSnp\n",
      "Table: snp141Common\n",
      "Table: snp141ExceptionDesc\n",
      "Table: snp141Flagged\n",
      "Table: snp141Mult\n",
      "Table: snp141OrthoPt4Pa2Rm3\n",
      "Table: snp141Seq\n",
      "Table: snp142\n",
      "Table: snp142CodingDbSnp\n",
      "Table: snp142Common\n",
      "Table: snp142ExceptionDesc\n",
      "Table: snp142Flagged\n",
      "Table: snp142Mult\n",
      "Table: snp142OrthoPt4Pa2Rm3\n",
      "Table: snp142Seq\n",
      "Table: snp144\n",
      "Table: snp144CodingDbSnp\n",
      "Table: snp144Common\n",
      "Table: snp144ExceptionDesc\n",
      "Table: snp144Flagged\n",
      "Table: snp144Mult\n",
      "Table: snp144OrthoPt4Pa2Rm3\n",
      "Table: snp144Seq\n",
      "Table: snp146\n",
      "Table: snp146CodingDbSnp\n",
      "Table: snp146Common\n",
      "Table: snp146ExceptionDesc\n",
      "Table: snp146Flagged\n",
      "Table: snp146Mult\n",
      "Table: snp146OrthoPt4Pa2Rm3\n",
      "Table: snp146Seq\n",
      "Table: snp147\n",
      "Table: snp147CodingDbSnp\n",
      "Table: snp147Common\n",
      "Table: snp147ExceptionDesc\n",
      "Table: snp147Flagged\n",
      "Table: snp147Mult\n",
      "Table: snp147OrthoPt4Pa2Rm3\n",
      "Table: snp147Seq\n",
      "Table: spAnnot\n",
      "Table: spMrna\n",
      "Table: spMut\n",
      "Table: spStruct\n",
      "Table: stsAlias\n",
      "Table: stsInfo2\n",
      "Table: stsMap\n",
      "Table: tRNAs\n",
      "Table: tableDescriptions\n",
      "Table: tableList\n",
      "Table: trackDb\n",
      "Table: ucscGenePfam\n",
      "Table: ucscRetroAli9\n",
      "Table: ucscRetroCds9\n",
      "Table: ucscRetroExtFile9\n",
      "Table: ucscRetroInfo9\n",
      "Table: ucscRetroOrtho9\n",
      "Table: ucscRetroSeq9\n",
      "Table: ucscScop\n",
      "Table: ucscToINSDC\n",
      "Table: wgEncodeBroadHistoneGm12878H3k27acStdSig\n",
      "Table: wgEncodeBroadHistoneGm12878H3k4me1StdSig\n",
      "Table: wgEncodeBroadHistoneGm12878H3k4me3StdSig\n",
      "Table: wgEncodeBroadHistoneH1hescH3k27acStdSig\n",
      "Table: wgEncodeBroadHistoneH1hescH3k4me1StdSig\n",
      "Table: wgEncodeBroadHistoneH1hescH3k4me3StdSig\n",
      "Table: wgEncodeBroadHistoneHsmmH3k27acStdSig\n",
      "Table: wgEncodeBroadHistoneHsmmH3k4me1StdSig\n",
      "Table: wgEncodeBroadHistoneHsmmH3k4me3StdSig\n",
      "Table: wgEncodeBroadHistoneHuvecH3k27acStdSig\n",
      "Table: wgEncodeBroadHistoneHuvecH3k4me1StdSig\n",
      "Table: wgEncodeBroadHistoneHuvecH3k4me3StdSig\n",
      "Table: wgEncodeBroadHistoneK562H3k27acStdSig\n",
      "Table: wgEncodeBroadHistoneK562H3k4me1StdSig\n",
      "Table: wgEncodeBroadHistoneK562H3k4me3StdSig\n",
      "Table: wgEncodeBroadHistoneNhekH3k27acStdSig\n",
      "Table: wgEncodeBroadHistoneNhekH3k4me1StdSig\n",
      "Table: wgEncodeBroadHistoneNhekH3k4me3StdSig\n",
      "Table: wgEncodeBroadHistoneNhlfH3k27acStdSig\n",
      "Table: wgEncodeBroadHistoneNhlfH3k4me1StdSig\n",
      "Table: wgEncodeBroadHistoneNhlfH3k4me3StdSig\n",
      "Table: wgEncodeCell\n",
      "Table: wgEncodeGencode2wayConsPseudoV20\n",
      "Table: wgEncodeGencode2wayConsPseudoV22\n",
      "Table: wgEncodeGencode2wayConsPseudoV23\n",
      "Table: wgEncodeGencode2wayConsPseudoV24\n",
      "Table: wgEncodeGencodeAnnotationRemarkV20\n",
      "Table: wgEncodeGencodeAnnotationRemarkV22\n",
      "Table: wgEncodeGencodeAnnotationRemarkV23\n",
      "Table: wgEncodeGencodeAnnotationRemarkV24\n",
      "Table: wgEncodeGencodeAttrsV20\n",
      "Table: wgEncodeGencodeAttrsV22\n",
      "Table: wgEncodeGencodeAttrsV23\n",
      "Table: wgEncodeGencodeAttrsV24\n",
      "Table: wgEncodeGencodeBasicV20\n",
      "Table: wgEncodeGencodeBasicV22\n",
      "Table: wgEncodeGencodeBasicV23\n",
      "Table: wgEncodeGencodeBasicV24\n",
      "Table: wgEncodeGencodeCompV20\n",
      "Table: wgEncodeGencodeCompV22\n",
      "Table: wgEncodeGencodeCompV23\n",
      "Table: wgEncodeGencodeCompV24\n",
      "Table: wgEncodeGencodeEntrezGeneV22\n",
      "Table: wgEncodeGencodeEntrezGeneV23\n",
      "Table: wgEncodeGencodeEntrezGeneV24\n",
      "Table: wgEncodeGencodeExonSupportV20\n",
      "Table: wgEncodeGencodeExonSupportV22\n",
      "Table: wgEncodeGencodeExonSupportV23\n",
      "Table: wgEncodeGencodeExonSupportV24\n",
      "Table: wgEncodeGencodeGeneSourceV20\n",
      "Table: wgEncodeGencodeGeneSourceV22\n",
      "Table: wgEncodeGencodeGeneSourceV23\n",
      "Table: wgEncodeGencodeGeneSourceV24\n",
      "Table: wgEncodeGencodePdbV20\n",
      "Table: wgEncodeGencodePdbV22\n",
      "Table: wgEncodeGencodePdbV23\n",
      "Table: wgEncodeGencodePdbV24\n",
      "Table: wgEncodeGencodePolyaV20\n",
      "Table: wgEncodeGencodePolyaV22\n",
      "Table: wgEncodeGencodePolyaV23\n",
      "Table: wgEncodeGencodePolyaV24\n",
      "Table: wgEncodeGencodePseudoGeneV20\n",
      "Table: wgEncodeGencodePseudoGeneV22\n",
      "Table: wgEncodeGencodePseudoGeneV23\n",
      "Table: wgEncodeGencodePseudoGeneV24\n",
      "Table: wgEncodeGencodePubMedV20\n",
      "Table: wgEncodeGencodePubMedV22\n",
      "Table: wgEncodeGencodePubMedV23\n",
      "Table: wgEncodeGencodePubMedV24\n",
      "Table: wgEncodeGencodeRefSeqV20\n",
      "Table: wgEncodeGencodeRefSeqV22\n",
      "Table: wgEncodeGencodeRefSeqV23\n",
      "Table: wgEncodeGencodeRefSeqV24\n",
      "Table: wgEncodeGencodeTagV20\n",
      "Table: wgEncodeGencodeTagV22\n",
      "Table: wgEncodeGencodeTagV23\n",
      "Table: wgEncodeGencodeTagV24\n",
      "Table: wgEncodeGencodeTranscriptSourceV20\n",
      "Table: wgEncodeGencodeTranscriptSourceV22\n",
      "Table: wgEncodeGencodeTranscriptSourceV23\n",
      "Table: wgEncodeGencodeTranscriptSourceV24\n",
      "Table: wgEncodeGencodeTranscriptSupportV20\n",
      "Table: wgEncodeGencodeTranscriptSupportV22\n",
      "Table: wgEncodeGencodeTranscriptSupportV23\n",
      "Table: wgEncodeGencodeTranscriptSupportV24\n",
      "Table: wgEncodeGencodeTranscriptionSupportLevelV20\n",
      "Table: wgEncodeGencodeTranscriptionSupportLevelV22\n",
      "Table: wgEncodeGencodeTranscriptionSupportLevelV23\n",
      "Table: wgEncodeGencodeTranscriptionSupportLevelV24\n",
      "Table: wgEncodeGencodeUniProtV20\n",
      "Table: wgEncodeGencodeUniProtV22\n",
      "Table: wgEncodeGencodeUniProtV23\n",
      "Table: wgEncodeGencodeUniProtV24\n",
      "Table: wgEncodeRegDnaseClustered\n",
      "Table: wgEncodeRegDnaseClusteredInputs\n",
      "Table: wgEncodeRegDnaseClusteredSources\n",
      "Table: wgEncodeRegDnaseUwA549Hotspot\n",
      "Table: wgEncodeRegDnaseUwA549Peak\n",
      "Table: wgEncodeRegDnaseUwA549Signal\n",
      "Table: wgEncodeRegDnaseUwAg04449Hotspot\n",
      "Table: wgEncodeRegDnaseUwAg04449Peak\n",
      "Table: wgEncodeRegDnaseUwAg04449Signal\n",
      "Table: wgEncodeRegDnaseUwAg04450Hotspot\n",
      "Table: wgEncodeRegDnaseUwAg04450Peak\n",
      "Table: wgEncodeRegDnaseUwAg04450Signal\n",
      "Table: wgEncodeRegDnaseUwAg09309Hotspot\n",
      "Table: wgEncodeRegDnaseUwAg09309Peak\n",
      "Table: wgEncodeRegDnaseUwAg09309Signal\n",
      "Table: wgEncodeRegDnaseUwAg09319Hotspot\n",
      "Table: wgEncodeRegDnaseUwAg09319Peak\n",
      "Table: wgEncodeRegDnaseUwAg09319Signal\n",
      "Table: wgEncodeRegDnaseUwAg10803Hotspot\n",
      "Table: wgEncodeRegDnaseUwAg10803Peak\n",
      "Table: wgEncodeRegDnaseUwAg10803Signal\n",
      "Table: wgEncodeRegDnaseUwAoafHotspot\n",
      "Table: wgEncodeRegDnaseUwAoafPeak\n",
      "Table: wgEncodeRegDnaseUwAoafSignal\n",
      "Table: wgEncodeRegDnaseUwBe2cHotspot\n",
      "Table: wgEncodeRegDnaseUwBe2cPeak\n",
      "Table: wgEncodeRegDnaseUwBe2cSignal\n",
      "Table: wgEncodeRegDnaseUwBjHotspot\n",
      "Table: wgEncodeRegDnaseUwBjPeak\n",
      "Table: wgEncodeRegDnaseUwBjSignal\n",
      "Table: wgEncodeRegDnaseUwBonemarrowmscHotspot\n",
      "Table: wgEncodeRegDnaseUwBonemarrowmscPeak\n",
      "Table: wgEncodeRegDnaseUwBonemarrowmscSignal\n",
      "Table: wgEncodeRegDnaseUwCaco2Hotspot\n",
      "Table: wgEncodeRegDnaseUwCaco2Peak\n",
      "Table: wgEncodeRegDnaseUwCaco2Signal\n",
      "Table: wgEncodeRegDnaseUwCd20ro01778Hotspot\n",
      "Table: wgEncodeRegDnaseUwCd20ro01778Peak\n",
      "Table: wgEncodeRegDnaseUwCd20ro01778Signal\n",
      "Table: wgEncodeRegDnaseUwGm04503Hotspot\n",
      "Table: wgEncodeRegDnaseUwGm04503Peak\n",
      "Table: wgEncodeRegDnaseUwGm04503Signal\n",
      "Table: wgEncodeRegDnaseUwGm04504Hotspot\n",
      "Table: wgEncodeRegDnaseUwGm04504Peak\n",
      "Table: wgEncodeRegDnaseUwGm04504Signal\n",
      "Table: wgEncodeRegDnaseUwGm06990Hotspot\n",
      "Table: wgEncodeRegDnaseUwGm06990Peak\n",
      "Table: wgEncodeRegDnaseUwGm06990Signal\n",
      "Table: wgEncodeRegDnaseUwGm12865Hotspot\n",
      "Table: wgEncodeRegDnaseUwGm12865Peak\n",
      "Table: wgEncodeRegDnaseUwGm12865Signal\n",
      "Table: wgEncodeRegDnaseUwGm12878Hotspot\n",
      "Table: wgEncodeRegDnaseUwGm12878Peak\n",
      "Table: wgEncodeRegDnaseUwGm12878Signal\n",
      "Table: wgEncodeRegDnaseUwH7hescDiffprota14dHotspot\n",
      "Table: wgEncodeRegDnaseUwH7hescDiffprota14dPeak\n",
      "Table: wgEncodeRegDnaseUwH7hescDiffprota14dSignal\n",
      "Table: wgEncodeRegDnaseUwH7hescDiffprota5dHotspot\n",
      "Table: wgEncodeRegDnaseUwH7hescDiffprota5dPeak\n",
      "Table: wgEncodeRegDnaseUwH7hescDiffprota5dSignal\n",
      "Table: wgEncodeRegDnaseUwH7hescHotspot\n",
      "Table: wgEncodeRegDnaseUwH7hescPeak\n",
      "Table: wgEncodeRegDnaseUwH7hescSignal\n",
      "Table: wgEncodeRegDnaseUwHacHotspot\n",
      "Table: wgEncodeRegDnaseUwHacPeak\n",
      "Table: wgEncodeRegDnaseUwHacSignal\n",
      "Table: wgEncodeRegDnaseUwHaepicHotspot\n",
      "Table: wgEncodeRegDnaseUwHaepicPeak\n",
      "Table: wgEncodeRegDnaseUwHaepicSignal\n",
      "Table: wgEncodeRegDnaseUwHahHotspot\n",
      "Table: wgEncodeRegDnaseUwHahPeak\n",
      "Table: wgEncodeRegDnaseUwHahSignal\n",
      "Table: wgEncodeRegDnaseUwHaspHotspot\n",
      "Table: wgEncodeRegDnaseUwHaspPeak\n",
      "Table: wgEncodeRegDnaseUwHaspSignal\n",
      "Table: wgEncodeRegDnaseUwHbmecHotspot\n",
      "Table: wgEncodeRegDnaseUwHbmecPeak\n",
      "Table: wgEncodeRegDnaseUwHbmecSignal\n",
      "Table: wgEncodeRegDnaseUwHbvsmcHotspot\n",
      "Table: wgEncodeRegDnaseUwHbvsmcPeak\n",
      "Table: wgEncodeRegDnaseUwHbvsmcSignal\n",
      "Table: wgEncodeRegDnaseUwHcfHotspot\n",
      "Table: wgEncodeRegDnaseUwHcfPeak\n",
      "Table: wgEncodeRegDnaseUwHcfSignal\n",
      "Table: wgEncodeRegDnaseUwHcfaaHotspot\n",
      "Table: wgEncodeRegDnaseUwHcfaaPeak\n",
      "Table: wgEncodeRegDnaseUwHcfaaSignal\n",
      "Table: wgEncodeRegDnaseUwHcmHotspot\n",
      "Table: wgEncodeRegDnaseUwHcmPeak\n",
      "Table: wgEncodeRegDnaseUwHcmSignal\n",
      "Table: wgEncodeRegDnaseUwHconfHotspot\n",
      "Table: wgEncodeRegDnaseUwHconfPeak\n",
      "Table: wgEncodeRegDnaseUwHconfSignal\n",
      "Table: wgEncodeRegDnaseUwHcpepicHotspot\n",
      "Table: wgEncodeRegDnaseUwHcpepicPeak\n",
      "Table: wgEncodeRegDnaseUwHcpepicSignal\n",
      "Table: wgEncodeRegDnaseUwHct116Hotspot\n",
      "Table: wgEncodeRegDnaseUwHct116Peak\n",
      "Table: wgEncodeRegDnaseUwHct116Signal\n",
      "Table: wgEncodeRegDnaseUwHeepicHotspot\n",
      "Table: wgEncodeRegDnaseUwHeepicPeak\n",
      "Table: wgEncodeRegDnaseUwHeepicSignal\n",
      "Table: wgEncodeRegDnaseUwHelas3Hotspot\n",
      "Table: wgEncodeRegDnaseUwHelas3Peak\n",
      "Table: wgEncodeRegDnaseUwHelas3Signal\n",
      "Table: wgEncodeRegDnaseUwHepg2Hotspot\n",
      "Table: wgEncodeRegDnaseUwHepg2Peak\n",
      "Table: wgEncodeRegDnaseUwHepg2Signal\n",
      "Table: wgEncodeRegDnaseUwHffHotspot\n",
      "Table: wgEncodeRegDnaseUwHffPeak\n",
      "Table: wgEncodeRegDnaseUwHffSignal\n",
      "Table: wgEncodeRegDnaseUwHffmycHotspot\n",
      "Table: wgEncodeRegDnaseUwHffmycPeak\n",
      "Table: wgEncodeRegDnaseUwHffmycSignal\n",
      "Table: wgEncodeRegDnaseUwHgfHotspot\n",
      "Table: wgEncodeRegDnaseUwHgfPeak\n",
      "Table: wgEncodeRegDnaseUwHgfSignal\n",
      "Table: wgEncodeRegDnaseUwHipepicHotspot\n",
      "Table: wgEncodeRegDnaseUwHipepicPeak\n",
      "Table: wgEncodeRegDnaseUwHipepicSignal\n",
      "Table: wgEncodeRegDnaseUwHl60Hotspot\n",
      "Table: wgEncodeRegDnaseUwHl60Peak\n",
      "Table: wgEncodeRegDnaseUwHl60Signal\n",
      "Table: wgEncodeRegDnaseUwHmecHotspot\n",
      "Table: wgEncodeRegDnaseUwHmecPeak\n",
      "Table: wgEncodeRegDnaseUwHmecSignal\n",
      "Table: wgEncodeRegDnaseUwHmfHotspot\n",
      "Table: wgEncodeRegDnaseUwHmfPeak\n",
      "Table: wgEncodeRegDnaseUwHmfSignal\n",
      "Table: wgEncodeRegDnaseUwHmvecdadHotspot\n",
      "Table: wgEncodeRegDnaseUwHmvecdadPeak\n",
      "Table: wgEncodeRegDnaseUwHmvecdadSignal\n",
      "Table: wgEncodeRegDnaseUwHmvecdbladHotspot\n",
      "Table: wgEncodeRegDnaseUwHmvecdbladPeak\n",
      "Table: wgEncodeRegDnaseUwHmvecdbladSignal\n",
      "Table: wgEncodeRegDnaseUwHmvecdblneoHotspot\n",
      "Table: wgEncodeRegDnaseUwHmvecdblneoPeak\n",
      "Table: wgEncodeRegDnaseUwHmvecdblneoSignal\n",
      "Table: wgEncodeRegDnaseUwHmvecdlyadHotspot\n",
      "Table: wgEncodeRegDnaseUwHmvecdlyadPeak\n",
      "Table: wgEncodeRegDnaseUwHmvecdlyadSignal\n",
      "Table: wgEncodeRegDnaseUwHmvecdlyneoHotspot\n",
      "Table: wgEncodeRegDnaseUwHmvecdlyneoPeak\n",
      "Table: wgEncodeRegDnaseUwHmvecdlyneoSignal\n",
      "Table: wgEncodeRegDnaseUwHmvecdneoHotspot\n",
      "Table: wgEncodeRegDnaseUwHmvecdneoPeak\n",
      "Table: wgEncodeRegDnaseUwHmvecdneoSignal\n",
      "Table: wgEncodeRegDnaseUwHmveclblHotspot\n",
      "Table: wgEncodeRegDnaseUwHmveclblPeak\n",
      "Table: wgEncodeRegDnaseUwHmveclblSignal\n",
      "Table: wgEncodeRegDnaseUwHmvecllyHotspot\n",
      "Table: wgEncodeRegDnaseUwHmvecllyPeak\n",
      "Table: wgEncodeRegDnaseUwHmvecllySignal\n",
      "Table: wgEncodeRegDnaseUwHnpcepicHotspot\n",
      "Table: wgEncodeRegDnaseUwHnpcepicPeak\n",
      "Table: wgEncodeRegDnaseUwHnpcepicSignal\n",
      "Table: wgEncodeRegDnaseUwHpafHotspot\n",
      "Table: wgEncodeRegDnaseUwHpafPeak\n",
      "Table: wgEncodeRegDnaseUwHpafSignal\n",
      "Table: wgEncodeRegDnaseUwHpdlfHotspot\n",
      "Table: wgEncodeRegDnaseUwHpdlfPeak\n",
      "Table: wgEncodeRegDnaseUwHpdlfSignal\n",
      "Table: wgEncodeRegDnaseUwHpfHotspot\n",
      "Table: wgEncodeRegDnaseUwHpfPeak\n",
      "Table: wgEncodeRegDnaseUwHpfSignal\n",
      "Table: wgEncodeRegDnaseUwHrcepicHotspot\n",
      "Table: wgEncodeRegDnaseUwHrcepicPeak\n",
      "Table: wgEncodeRegDnaseUwHrcepicSignal\n",
      "Table: wgEncodeRegDnaseUwHreHotspot\n",
      "Table: wgEncodeRegDnaseUwHrePeak\n",
      "Table: wgEncodeRegDnaseUwHreSignal\n",
      "Table: wgEncodeRegDnaseUwHrgecHotspot\n",
      "Table: wgEncodeRegDnaseUwHrgecPeak\n",
      "Table: wgEncodeRegDnaseUwHrgecSignal\n",
      "Table: wgEncodeRegDnaseUwHrpepicHotspot\n",
      "Table: wgEncodeRegDnaseUwHrpepicPeak\n",
      "Table: wgEncodeRegDnaseUwHrpepicSignal\n",
      "Table: wgEncodeRegDnaseUwHsmmHotspot\n",
      "Table: wgEncodeRegDnaseUwHsmmPeak\n",
      "Table: wgEncodeRegDnaseUwHsmmSignal\n",
      "Table: wgEncodeRegDnaseUwHsmmtubeHotspot\n",
      "Table: wgEncodeRegDnaseUwHsmmtubePeak\n",
      "Table: wgEncodeRegDnaseUwHsmmtubeSignal\n",
      "Table: wgEncodeRegDnaseUwHuvecHotspot\n",
      "Table: wgEncodeRegDnaseUwHuvecPeak\n",
      "Table: wgEncodeRegDnaseUwHuvecSignal\n",
      "Table: wgEncodeRegDnaseUwHvmfHotspot\n",
      "Table: wgEncodeRegDnaseUwHvmfPeak\n",
      "Table: wgEncodeRegDnaseUwHvmfSignal\n",
      "Table: wgEncodeRegDnaseUwJurkatHotspot\n",
      "Table: wgEncodeRegDnaseUwJurkatPeak\n",
      "Table: wgEncodeRegDnaseUwJurkatSignal\n",
      "Table: wgEncodeRegDnaseUwK562Hotspot\n",
      "Table: wgEncodeRegDnaseUwK562Peak\n",
      "Table: wgEncodeRegDnaseUwK562Signal\n",
      "Table: wgEncodeRegDnaseUwLhcnm2Diff4dHotspot\n",
      "Table: wgEncodeRegDnaseUwLhcnm2Diff4dPeak\n",
      "Table: wgEncodeRegDnaseUwLhcnm2Diff4dSignal\n",
      "Table: wgEncodeRegDnaseUwLhcnm2Hotspot\n",
      "Table: wgEncodeRegDnaseUwLhcnm2Peak\n",
      "Table: wgEncodeRegDnaseUwLhcnm2Signal\n",
      "Table: wgEncodeRegDnaseUwLncapHotspot\n",
      "Table: wgEncodeRegDnaseUwLncapPeak\n",
      "Table: wgEncodeRegDnaseUwLncapSignal\n",
      "Table: wgEncodeRegDnaseUwM059jHotspot\n",
      "Table: wgEncodeRegDnaseUwM059jPeak\n",
      "Table: wgEncodeRegDnaseUwM059jSignal\n",
      "Table: wgEncodeRegDnaseUwMcf7Estradiol100nm1hrHotspot\n",
      "Table: wgEncodeRegDnaseUwMcf7Estradiol100nm1hrPeak\n",
      "Table: wgEncodeRegDnaseUwMcf7Estradiol100nm1hrSignal\n",
      "Table: wgEncodeRegDnaseUwMcf7Estradiolctrl0hrHotspot\n",
      "Table: wgEncodeRegDnaseUwMcf7Estradiolctrl0hrPeak\n",
      "Table: wgEncodeRegDnaseUwMcf7Estradiolctrl0hrSignal\n",
      "Table: wgEncodeRegDnaseUwMcf7Hotspot\n",
      "Table: wgEncodeRegDnaseUwMcf7Peak\n",
      "Table: wgEncodeRegDnaseUwMcf7Signal\n",
      "Table: wgEncodeRegDnaseUwMonocytescd14ro01746Hotspot\n",
      "Table: wgEncodeRegDnaseUwMonocytescd14ro01746Peak\n",
      "Table: wgEncodeRegDnaseUwMonocytescd14ro01746Signal\n",
      "Table: wgEncodeRegDnaseUwNb4Hotspot\n",
      "Table: wgEncodeRegDnaseUwNb4Peak\n",
      "Table: wgEncodeRegDnaseUwNb4Signal\n",
      "Table: wgEncodeRegDnaseUwNhaHotspot\n",
      "Table: wgEncodeRegDnaseUwNhaPeak\n",
      "Table: wgEncodeRegDnaseUwNhaSignal\n",
      "Table: wgEncodeRegDnaseUwNhberaHotspot\n",
      "Table: wgEncodeRegDnaseUwNhberaPeak\n",
      "Table: wgEncodeRegDnaseUwNhberaSignal\n",
      "Table: wgEncodeRegDnaseUwNhdfadHotspot\n",
      "Table: wgEncodeRegDnaseUwNhdfadPeak\n",
      "Table: wgEncodeRegDnaseUwNhdfadSignal\n",
      "Table: wgEncodeRegDnaseUwNhdfneoHotspot\n",
      "Table: wgEncodeRegDnaseUwNhdfneoPeak\n",
      "Table: wgEncodeRegDnaseUwNhdfneoSignal\n",
      "Table: wgEncodeRegDnaseUwNhekHotspot\n",
      "Table: wgEncodeRegDnaseUwNhekPeak\n",
      "Table: wgEncodeRegDnaseUwNhekSignal\n",
      "Table: wgEncodeRegDnaseUwNhlfHotspot\n",
      "Table: wgEncodeRegDnaseUwNhlfPeak\n",
      "Table: wgEncodeRegDnaseUwNhlfSignal\n",
      "Table: wgEncodeRegDnaseUwNt2d1Hotspot\n",
      "Table: wgEncodeRegDnaseUwNt2d1Peak\n",
      "Table: wgEncodeRegDnaseUwNt2d1Signal\n",
      "Table: wgEncodeRegDnaseUwPanc1Hotspot\n",
      "Table: wgEncodeRegDnaseUwPanc1Peak\n",
      "Table: wgEncodeRegDnaseUwPanc1Signal\n",
      "Table: wgEncodeRegDnaseUwPrecHotspot\n",
      "Table: wgEncodeRegDnaseUwPrecPeak\n",
      "Table: wgEncodeRegDnaseUwPrecSignal\n",
      "Table: wgEncodeRegDnaseUwRpmi7951Hotspot\n",
      "Table: wgEncodeRegDnaseUwRpmi7951Peak\n",
      "Table: wgEncodeRegDnaseUwRpmi7951Signal\n",
      "Table: wgEncodeRegDnaseUwRptecHotspot\n",
      "Table: wgEncodeRegDnaseUwRptecPeak\n",
      "Table: wgEncodeRegDnaseUwRptecSignal\n",
      "Table: wgEncodeRegDnaseUwSaecHotspot\n",
      "Table: wgEncodeRegDnaseUwSaecPeak\n",
      "Table: wgEncodeRegDnaseUwSaecSignal\n",
      "Table: wgEncodeRegDnaseUwSkmcHotspot\n",
      "Table: wgEncodeRegDnaseUwSkmcPeak\n",
      "Table: wgEncodeRegDnaseUwSkmcSignal\n",
      "Table: wgEncodeRegDnaseUwSknmcHotspot\n",
      "Table: wgEncodeRegDnaseUwSknmcPeak\n",
      "Table: wgEncodeRegDnaseUwSknmcSignal\n",
      "Table: wgEncodeRegDnaseUwSknshraHotspot\n",
      "Table: wgEncodeRegDnaseUwSknshraPeak\n",
      "Table: wgEncodeRegDnaseUwSknshraSignal\n",
      "Table: wgEncodeRegDnaseUwT47dHotspot\n",
      "Table: wgEncodeRegDnaseUwT47dPeak\n",
      "Table: wgEncodeRegDnaseUwT47dSignal\n",
      "Table: wgEncodeRegDnaseUwTh1Hotspot\n",
      "Table: wgEncodeRegDnaseUwTh1Peak\n",
      "Table: wgEncodeRegDnaseUwTh1Signal\n",
      "Table: wgEncodeRegDnaseUwTh1wb54553204Hotspot\n",
      "Table: wgEncodeRegDnaseUwTh1wb54553204Peak\n",
      "Table: wgEncodeRegDnaseUwTh1wb54553204Signal\n",
      "Table: wgEncodeRegDnaseUwTh2Hotspot\n",
      "Table: wgEncodeRegDnaseUwTh2Peak\n",
      "Table: wgEncodeRegDnaseUwTh2Signal\n",
      "Table: wgEncodeRegDnaseUwWerirb1Hotspot\n",
      "Table: wgEncodeRegDnaseUwWerirb1Peak\n",
      "Table: wgEncodeRegDnaseUwWerirb1Signal\n",
      "Table: wgEncodeRegDnaseUwWi384ohtam20nm72hrHotspot\n",
      "Table: wgEncodeRegDnaseUwWi384ohtam20nm72hrPeak\n",
      "Table: wgEncodeRegDnaseUwWi384ohtam20nm72hrSignal\n",
      "Table: wgEncodeRegDnaseUwWi38Hotspot\n",
      "Table: wgEncodeRegDnaseUwWi38Peak\n",
      "Table: wgEncodeRegDnaseUwWi38Signal\n",
      "Table: wgEncodeRegTxnCaltechRnaSeqGm12878R2x75Il200SigPooled\n",
      "Table: wgEncodeRegTxnCaltechRnaSeqH1hescR2x75Il200SigPooled\n",
      "Table: wgEncodeRegTxnCaltechRnaSeqHelas3R2x75Il200SigPooled\n",
      "Table: wgEncodeRegTxnCaltechRnaSeqHepg2R2x75Il200SigPooled\n",
      "Table: wgEncodeRegTxnCaltechRnaSeqHsmmR2x75Il200SigPooled\n",
      "Table: wgEncodeRegTxnCaltechRnaSeqHuvecR2x75Il200SigPooled\n",
      "Table: wgEncodeRegTxnCaltechRnaSeqK562R2x75Il200SigPooled\n",
      "Table: wgEncodeRegTxnCaltechRnaSeqNhekR2x75Il200SigPooled\n",
      "Table: wgEncodeRegTxnCaltechRnaSeqNhlfR2x75Il200SigPooled\n",
      "Table: wgEncodeTreatment\n",
      "Table: wgRna\n",
      "Table: windowmaskerSdust\n",
      "Table: xenoEst\n",
      "Table: xenoMrna\n",
      "Table: xenoRefFlat\n",
      "Table: xenoRefGene\n",
      "Table: xenoRefSeqAli\n"
     ]
    }
   ],
   "source": [
    "connection = engine.connect()\n",
    "result = connection.execute(\"SHOW TABLES\")\n",
    "for row in result:\n",
    "    print(\"Table:\", row[0])\n",
    "connection.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Columns: ('bin', 'smallint(5) unsigned', 'NO', '', None, '')\n",
      "Columns: ('name', 'varchar(255)', 'NO', 'MUL', None, '')\n",
      "Columns: ('chrom', 'varchar(255)', 'NO', 'MUL', None, '')\n",
      "Columns: ('strand', 'char(1)', 'NO', '', None, '')\n",
      "Columns: ('txStart', 'int(10) unsigned', 'NO', '', None, '')\n",
      "Columns: ('txEnd', 'int(10) unsigned', 'NO', '', None, '')\n",
      "Columns: ('cdsStart', 'int(10) unsigned', 'NO', '', None, '')\n",
      "Columns: ('cdsEnd', 'int(10) unsigned', 'NO', '', None, '')\n",
      "Columns: ('exonCount', 'int(10) unsigned', 'NO', '', None, '')\n",
      "Columns: ('exonStarts', 'longblob', 'NO', '', None, '')\n",
      "Columns: ('exonEnds', 'longblob', 'NO', '', None, '')\n",
      "Columns: ('score', 'int(11)', 'YES', '', None, '')\n",
      "Columns: ('name2', 'varchar(255)', 'NO', 'MUL', None, '')\n",
      "Columns: ('cdsStartStat', \"enum('none','unk','incmpl','cmpl')\", 'NO', '', None, '')\n",
      "Columns: ('cdsEndStat', \"enum('none','unk','incmpl','cmpl')\", 'NO', '', None, '')\n",
      "Columns: ('exonFrames', 'longblob', 'NO', '', None, '')\n"
     ]
    }
   ],
   "source": [
    "# Connection supports context manager\n",
    "with engine.connect() as connection:\n",
    "    result = connection.execute(\"DESCRIBE refGene\")\n",
    "    for row in result:\n",
    "        print(\"Columns:\", row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record # 0\n",
      "(585, 'NR_046018', 'chr1', '+', 11873, 14409, 14409, 14409, 3, b'11873,12612,13220,', b'12227,12721,14409,', 0, 'DDX11L1', 'unk', 'unk', b'-1,-1,-1,')\n",
      "Record # 1\n",
      "(585, 'NR_024540', 'chr1', '-', 14361, 29370, 29370, 29370, 11, b'14361,14969,15795,16606,16857,17232,17605,17914,18267,24737,29320,', b'14829,15038,15947,16765,17055,17368,17742,18061,18366,24891,29370,', 0, 'WASH7P', 'unk', 'unk', b'-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,')\n",
      "Record # 2\n",
      "(1078, 'NR_104148', 'chr7', '+', 64666082, 64687830, 64687830, 64687830, 4, b'64666082,64669036,64679176,64684334,', b'64666285,64669178,64679336,64687830,', 0, 'ZNF107', 'unk', 'unk', b'-1,-1,-1,-1,')\n",
      "Record # 3\n",
      "(14, 'NR_111960', 'chrX', '+', 44873174, 45112612, 45112612, 45112612, 28, b'44873174,44873923,44961283,44974665,45010960,45020609,45034930,45037654,45051708,45053828,45059005,45059246,45060608,45061323,45063421,45069578,45076696,45078399,45079145,45082575,45082714,45083459,45085864,45089742,45090722,45107409,45110078,45111381,', b'44873712,44873987,44961392,44974715,45011019,45020730,45034985,45037689,45051802,45053955,45059104,45059466,45060764,45061419,45063817,45070357,45076826,45078505,45079351,45082640,45082789,45083608,45085979,45089930,45090832,45107536,45110249,45112612,', 0, 'KDM6A', 'unk', 'unk', b'-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,-1,')\n",
      "Record # 4\n",
      "(585, 'NR_026818', 'chr1', '-', 34610, 36081, 36081, 36081, 3, b'34610,35276,35720,', b'35174,35481,36081,', 0, 'FAM138A', 'unk', 'unk', b'-1,-1,-1,')\n"
     ]
    }
   ],
   "source": [
    "# Engine can automatically take a connection from the pool\n",
    "# SQL query - fetch all columns from the first 5 records\n",
    "result = engine.execute(\"SELECT * FROM refGene LIMIT 5\")\n",
    "for i, row in enumerate(result):\n",
    "    print(\"Record #\", i)\n",
    "    pprint(row)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record # 0\n",
      "\tGene NR_046018 (DDX11L1)\n",
      "\tCDS location chr1 14409-14409 on strand +\n",
      "Record # 1\n",
      "\tGene NR_024540 (WASH7P)\n",
      "\tCDS location chr1 29370-29370 on strand -\n",
      "Record # 2\n",
      "\tGene NR_104148 (ZNF107)\n",
      "\tCDS location chr7 64687830-64687830 on strand +\n",
      "Record # 3\n",
      "\tGene NR_111960 (KDM6A)\n",
      "\tCDS location chrX 45112612-45112612 on strand +\n",
      "Record # 4\n",
      "\tGene NR_026818 (FAM138A)\n",
      "\tCDS location chr1 36081-36081 on strand -\n"
     ]
    }
   ],
   "source": [
    "# Selected columns\n",
    "with engine.connect() as connection:\n",
    "    result = connection.execute(\"SELECT name, name2, chrom, strand, cdsStart, cdsEnd FROM refGene LIMIT 5\")\n",
    "    for i, row in enumerate(result):\n",
    "        print(\"Record #\", i)\n",
    "        print(\"\\tGene {} ({})\".format(row['name'], row['name2']))\n",
    "        print(\"\\tCDS location {} {}-{} on strand {}\".format(row['chrom'], row['cdsStart'], row['cdsEnd'], row['strand']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Record # 0\n",
      "\tGene NR_104148 (ZNF107)\n",
      "\tCDS location chr7 64687830-64687830 on strand +\n",
      "Record # 1\n",
      "\tGene NM_016220 (ZNF107)\n",
      "\tCDS location chr7 64691941-64708656 on strand +\n",
      "Record # 2\n",
      "\tGene NR_104149 (ZNF107)\n",
      "\tCDS location chr7 64687830-64687830 on strand +\n",
      "Record # 3\n",
      "\tGene NM_001282360 (ZNF107)\n",
      "\tCDS location chr7 64666282-64708656 on strand +\n",
      "Record # 4\n",
      "\tGene NM_001282359 (ZNF107)\n",
      "\tCDS location chr7 64666282-64708656 on strand +\n",
      "Record # 5\n",
      "\tGene NM_001013746 (ZNF107)\n",
      "\tCDS location chr7 64691941-64708656 on strand +\n"
     ]
    }
   ],
   "source": [
    "# Selected columns and rows using SQL\n",
    "with engine.connect() as connection:\n",
    "    result = connection.execute(\"\"\"\n",
    "        SELECT\n",
    "            name, name2, chrom, strand, cdsStart, cdsEnd\n",
    "        FROM\n",
    "            refGene\n",
    "        WHERE\n",
    "            name2='ZNF107'\n",
    "        \"\"\")\n",
    "    for i, row in enumerate(result):\n",
    "        print(\"Record #\", i)\n",
    "        print(\"\\tGene {} ({})\".format(row['name'], row['name2']))\n",
    "        print(\"\\tCDS location {} {}-{} on strand {}\".format(row['chrom'], row['cdsStart'], row['cdsEnd'], row['strand']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['bin',\n",
      " 'name',\n",
      " 'chrom',\n",
      " 'strand',\n",
      " 'txStart',\n",
      " 'txEnd',\n",
      " 'cdsStart',\n",
      " 'cdsEnd',\n",
      " 'exonCount',\n",
      " 'exonStarts',\n",
      " 'exonEnds',\n",
      " 'score',\n",
      " 'name2',\n",
      " 'cdsStartStat',\n",
      " 'cdsEndStat',\n",
      " 'exonFrames']\n",
      "\n",
      "strand CHAR(1)\n"
     ]
    }
   ],
   "source": [
    "meta = MetaData(bind=engine)\n",
    "\n",
    "# Usually we can then produce a set of mappings from this MetaData.\n",
    "# Base = automap_base(bind=e, metadata=meta)\n",
    "\n",
    "# calling prepare() just sets up mapped classes and relationships.\n",
    "# Base.prepare()\n",
    "\n",
    "# Then, the classes are ready\n",
    "# Base.classes.refGene\n",
    "\n",
    "# Unfortunately, not in this case - there are no primary keys and automap_base requires them\n",
    "\n",
    "# Alternatively, we can use:\n",
    "meta.reflect(only=['refGene', 'snp147Common'])\n",
    "\n",
    "# knownGeneMrnahg38\n",
    "\n",
    "# However, we need to modify metadata and add a primary key:\n",
    "gene_table = Table('refGene',\n",
    "    meta,\n",
    "    PrimaryKeyConstraint('name'),\n",
    "    extend_existing=True)\n",
    "\n",
    "pprint(gene_table.columns.keys())\n",
    "print()\n",
    "print(gene_table.c.strand.name, gene_table.c.strand.type)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import sqlalchemy.orm as orm\n",
    "session = orm.sessionmaker()(bind=engine)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<__main__.Gene object at 0x105509160>\n"
     ]
    }
   ],
   "source": [
    "class Gene(object):\n",
    "    pass\n",
    "\n",
    "orm.mapper(Gene, gene_table)\n",
    "\n",
    "g = session.query(Gene).filter_by(name2='ZNF107').first()\n",
    "print(g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene NR_104148 (ZNF107)\n",
      "CDS location chr7 64687830-64687830 on strand +\n",
      "\n",
      "_sa_instance_state  : <sqlalchemy.orm.state.InstanceState object at 0x10551e1d0>\n",
      "bin                 : 1078\n",
      "cdsEnd              : 64687830\n",
      "cdsEndStat          : unk\n",
      "cdsStart            : 64687830\n",
      "cdsStartStat        : unk\n",
      "chrom               : chr7\n",
      "exonCount           : 4\n",
      "exonEnds            : b'64666285,64669178,64679336,64687830,'\n",
      "exonFrames          : b'-1,-1,-1,-1,'\n",
      "exonStarts          : b'64666082,64669036,64679176,64684334,'\n",
      "name                : NR_104148\n",
      "name2               : ZNF107\n",
      "score               : 0\n",
      "strand              : +\n",
      "txEnd               : 64687830\n",
      "txStart             : 64666082\n"
     ]
    }
   ],
   "source": [
    "class DatabaseObject(object):\n",
    "    def __repr__(self):\n",
    "        return \"\\n\".join(\n",
    "            [\"{:20s}: {}\".format(key, self.__dict__[key]) for key in sorted(self.__dict__.keys())]\n",
    "        )\n",
    "\n",
    "\n",
    "class Gene(DatabaseObject):\n",
    "    pass\n",
    "\n",
    "\n",
    "orm.mapper(Gene, gene_table)\n",
    "\n",
    "g = session.query(Gene).filter_by(name2='ZNF107').first()\n",
    "print(\"Gene {} ({})\".format(g.name, g.name2))\n",
    "print(\"CDS location {} {}-{} on strand {}\".format(g.chrom, g.cdsStart, g.cdsEnd, g.strand))\n",
    "print()\n",
    "print(g)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene NR_104148 (ZNF107)\n",
      "CDS location chr7 64687830-64687830 on strand +\n"
     ]
    }
   ],
   "source": [
    "class Gene(DatabaseObject):\n",
    "    def __repr__(self):\n",
    "        return(\"Gene {} ({})\\nCDS location {} {}-{} on strand {}\".format(\n",
    "            g.name, g.name2, g.chrom, g.cdsStart, g.cdsEnd, g.strand))\n",
    "\n",
    "orm.mapper(Gene, gene_table)\n",
    "\n",
    "print(session.query(Gene).filter_by(name2='ZNF107').first())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('bin', 'smallint(5) unsigned', 'NO', '', None, '')\n",
      "('chrom', 'varchar(31)', 'NO', 'MUL', None, '')\n",
      "('chromStart', 'int(10) unsigned', 'NO', '', None, '')\n",
      "('chromEnd', 'int(10) unsigned', 'NO', '', None, '')\n",
      "('name', 'varchar(15)', 'NO', 'MUL', None, '')\n",
      "('score', 'smallint(5) unsigned', 'NO', '', None, '')\n",
      "('strand', \"enum('+','-')\", 'NO', '', None, '')\n",
      "('refNCBI', 'blob', 'NO', '', None, '')\n",
      "('refUCSC', 'blob', 'NO', '', None, '')\n",
      "('observed', 'varchar(255)', 'NO', '', None, '')\n",
      "('molType', \"enum('unknown','genomic','cDNA')\", 'NO', '', None, '')\n",
      "('class', \"enum('single','in-del','microsatellite','named','mnp','insertion','deletion')\", 'NO', '', None, '')\n",
      "('valid', \"set('unknown','by-cluster','by-frequency','by-submitter','by-2hit-2allele','by-hapmap','by-1000genomes')\", 'NO', '', None, '')\n",
      "('avHet', 'float', 'NO', '', None, '')\n",
      "('avHetSE', 'float', 'NO', '', None, '')\n",
      "('func', \"set('unknown','coding-synon','intron','near-gene-3','near-gene-5','ncRNA','nonsense','missense','stop-loss','frameshift','cds-indel','untranslated-3','untranslated-5','splice-3','splice-5')\", 'NO', '', None, '')\n",
      "('locType', \"enum('range','exact','between','rangeInsertion','rangeSubstitution','rangeDeletion','fuzzy')\", 'NO', '', None, '')\n",
      "('weight', 'int(10) unsigned', 'NO', '', None, '')\n",
      "('exceptions', \"set('RefAlleleMismatch','DuplicateObserved','MixedObserved','FlankMismatchGenomeLonger','FlankMismatchGenomeEqual','FlankMismatchGenomeShorter','SingleClassLongerSpan','SingleClassZeroSpan','SingleClassTriAllelic','SingleClassQuadAllelic','ObservedWrongFormat','ObservedTooLong','ObservedContainsIupac','ObservedMismatch','MultipleAlignments','NonIntegerChromCount','AlleleFreqSumNot1','SingleAlleleFreq','InconsistentAlleles')\", 'NO', '', None, '')\n",
      "('submitterCount', 'smallint(5) unsigned', 'NO', '', None, '')\n",
      "('submitters', 'longblob', 'NO', '', None, '')\n",
      "('alleleFreqCount', 'smallint(5) unsigned', 'NO', '', None, '')\n",
      "('alleles', 'longblob', 'NO', '', None, '')\n",
      "('alleleNs', 'longblob', 'NO', '', None, '')\n",
      "('alleleFreqs', 'longblob', 'NO', '', None, '')\n",
      "('bitfields', \"set('clinically-assoc','maf-5-some-pop','maf-5-all-pops','has-omim-omia','microattr-tpa','submitted-by-lsdb','genotype-conflict','rs-cluster-nonoverlapping-alleles','observed-mismatch')\", 'NO', '', None, '')\n"
     ]
    }
   ],
   "source": [
    "print(\"\\n\".join(map(repr, engine.execute(\"DESCRIBE snp147Common\"))))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "_sa_instance_state  : <sqlalchemy.orm.state.InstanceState object at 0x105582358>\n",
      "alleleFreqCount     : 2\n",
      "alleleFreqs         : b'0.574681,0.425319,'\n",
      "alleleNs            : b'2878.000000,2130.000000,'\n",
      "alleles             : b'-,C,'\n",
      "avHet               : 0.488846\n",
      "avHetSE             : 0.073843\n",
      "bin                 : 585\n",
      "bitfields           : {'maf-5-some-pop', 'maf-5-all-pops'}\n",
      "chrom               : chr1\n",
      "chromEnd            : 10177\n",
      "chromStart          : 10177\n",
      "class               : insertion\n",
      "exceptions          : set()\n",
      "func                : {'near-gene-5'}\n",
      "locType             : between\n",
      "molType             : genomic\n",
      "name                : rs367896724\n",
      "observed            : -/C\n",
      "refNCBI             : b'-'\n",
      "refUCSC             : b'-'\n",
      "score               : 0\n",
      "strand              : +\n",
      "submitterCount      : 2\n",
      "submitters          : b'1000GENOMES,BILGI_BIOE,'\n",
      "valid               : {'by-frequency', 'by-cluster', 'by-1000genomes'}\n",
      "weight              : 1\n"
     ]
    }
   ],
   "source": [
    "snp_table = Table('snp147Common',\n",
    "    meta,\n",
    "    PrimaryKeyConstraint('name'),\n",
    "    extend_existing=True)\n",
    "\n",
    "# class SNP(object): pass\n",
    "class SNP(DatabaseObject):\n",
    "    snp_class = Column('class', String)\n",
    "\n",
    "orm.mapper(SNP, snp_table)\n",
    "\n",
    "print(session.query(SNP).first())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "SELECT `snp147Common`.bin, `snp147Common`.chrom, `snp147Common`.`chromStart`, `snp147Common`.`chromEnd`, `snp147Common`.name, `snp147Common`.score, `snp147Common`.strand, `snp147Common`.`refNCBI`, `snp147Common`.`refUCSC`, `snp147Common`.observed, `snp147Common`.`molType`, `snp147Common`.class, `snp147Common`.valid, `snp147Common`.`avHet`, `snp147Common`.`avHetSE`, `snp147Common`.func, `snp147Common`.`locType`, `snp147Common`.weight, `snp147Common`.exceptions, `snp147Common`.`submitterCount`, `snp147Common`.submitters, `snp147Common`.`alleleFreqCount`, `snp147Common`.alleles, `snp147Common`.`alleleNs`, `snp147Common`.`alleleFreqs`, `snp147Common`.bitfields \n",
      "FROM `snp147Common` \n",
      "WHERE `snp147Common`.chrom = %(chrom_1)s \n",
      " LIMIT %(param_1)s\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>bin</th>\n",
       "      <th>chrom</th>\n",
       "      <th>chromStart</th>\n",
       "      <th>chromEnd</th>\n",
       "      <th>name</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>refNCBI</th>\n",
       "      <th>refUCSC</th>\n",
       "      <th>observed</th>\n",
       "      <th>...</th>\n",
       "      <th>locType</th>\n",
       "      <th>weight</th>\n",
       "      <th>exceptions</th>\n",
       "      <th>submitterCount</th>\n",
       "      <th>submitters</th>\n",
       "      <th>alleleFreqCount</th>\n",
       "      <th>alleles</th>\n",
       "      <th>alleleNs</th>\n",
       "      <th>alleleFreqs</th>\n",
       "      <th>bitfields</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>585</td>\n",
       "      <td>chrY</td>\n",
       "      <td>10020</td>\n",
       "      <td>10020</td>\n",
       "      <td>rs745593600</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>b'-'</td>\n",
       "      <td>b'-'</td>\n",
       "      <td>-/A/AAC</td>\n",
       "      <td>...</td>\n",
       "      <td>between</td>\n",
       "      <td>1</td>\n",
       "      <td>{MixedObserved}</td>\n",
       "      <td>1</td>\n",
       "      <td>b'1000GENOMES,'</td>\n",
       "      <td>3</td>\n",
       "      <td>b'-,A,AAC,'</td>\n",
       "      <td>b'4906.000000,10.000000,92.000000,'</td>\n",
       "      <td>b'0.979633,0.001997,0.018371,'</td>\n",
       "      <td>{}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>585</td>\n",
       "      <td>chrY</td>\n",
       "      <td>10034</td>\n",
       "      <td>10036</td>\n",
       "      <td>rs201278642</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>b'CC'</td>\n",
       "      <td>b'CC'</td>\n",
       "      <td>-/CC</td>\n",
       "      <td>...</td>\n",
       "      <td>range</td>\n",
       "      <td>1</td>\n",
       "      <td>{MixedObserved}</td>\n",
       "      <td>2</td>\n",
       "      <td>b'1000GENOMES,SSMP,'</td>\n",
       "      <td>2</td>\n",
       "      <td>b'-,CC,'</td>\n",
       "      <td>b'369.000000,4637.000000,'</td>\n",
       "      <td>b'0.073711,0.926288,'</td>\n",
       "      <td>{maf-5-some-pop}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>585</td>\n",
       "      <td>chrY</td>\n",
       "      <td>10051</td>\n",
       "      <td>10052</td>\n",
       "      <td>rs186434315</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>b'T'</td>\n",
       "      <td>b'T'</td>\n",
       "      <td>A/T</td>\n",
       "      <td>...</td>\n",
       "      <td>exact</td>\n",
       "      <td>1</td>\n",
       "      <td>{}</td>\n",
       "      <td>2</td>\n",
       "      <td>b'1000GENOMES,SSMP,'</td>\n",
       "      <td>2</td>\n",
       "      <td>b'A,T,'</td>\n",
       "      <td>b'1582.000000,3426.000000,'</td>\n",
       "      <td>b'0.315895,0.684105,'</td>\n",
       "      <td>{maf-5-some-pop, maf-5-all-pops}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>585</td>\n",
       "      <td>chrY</td>\n",
       "      <td>10059</td>\n",
       "      <td>10060</td>\n",
       "      <td>rs148832940</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>b'G'</td>\n",
       "      <td>b'G'</td>\n",
       "      <td>C/G</td>\n",
       "      <td>...</td>\n",
       "      <td>exact</td>\n",
       "      <td>1</td>\n",
       "      <td>{}</td>\n",
       "      <td>2</td>\n",
       "      <td>b'1000GENOMES,WEILL_CORNELL_DGM,'</td>\n",
       "      <td>2</td>\n",
       "      <td>b'C,G,'</td>\n",
       "      <td>b'245.000000,4763.000000,'</td>\n",
       "      <td>b'0.048922,0.951078,'</td>\n",
       "      <td>{maf-5-some-pop}</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>585</td>\n",
       "      <td>chrY</td>\n",
       "      <td>10068</td>\n",
       "      <td>10069</td>\n",
       "      <td>rs111065272</td>\n",
       "      <td>0</td>\n",
       "      <td>+</td>\n",
       "      <td>b'T'</td>\n",
       "      <td>b'T'</td>\n",
       "      <td>A/T</td>\n",
       "      <td>...</td>\n",
       "      <td>exact</td>\n",
       "      <td>1</td>\n",
       "      <td>{}</td>\n",
       "      <td>4</td>\n",
       "      <td>b'1000GENOMES,COMPLETE_GENOMICS,SSMP,WEILL_COR...</td>\n",
       "      <td>2</td>\n",
       "      <td>b'A,T,'</td>\n",
       "      <td>b'1585.000000,3423.000000,'</td>\n",
       "      <td>b'0.316494,0.683506,'</td>\n",
       "      <td>{maf-5-some-pop, maf-5-all-pops}</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 26 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   bin chrom  chromStart  chromEnd         name  score strand refNCBI refUCSC  \\\n",
       "0  585  chrY       10020     10020  rs745593600      0      +    b'-'    b'-'   \n",
       "1  585  chrY       10034     10036  rs201278642      0      +   b'CC'   b'CC'   \n",
       "2  585  chrY       10051     10052  rs186434315      0      +    b'T'    b'T'   \n",
       "3  585  chrY       10059     10060  rs148832940      0      +    b'G'    b'G'   \n",
       "4  585  chrY       10068     10069  rs111065272      0      +    b'T'    b'T'   \n",
       "\n",
       "  observed                ...                 locType weight       exceptions  \\\n",
       "0  -/A/AAC                ...                 between      1  {MixedObserved}   \n",
       "1     -/CC                ...                   range      1  {MixedObserved}   \n",
       "2      A/T                ...                   exact      1               {}   \n",
       "3      C/G                ...                   exact      1               {}   \n",
       "4      A/T                ...                   exact      1               {}   \n",
       "\n",
       "   submitterCount                                         submitters  \\\n",
       "0               1                                    b'1000GENOMES,'   \n",
       "1               2                               b'1000GENOMES,SSMP,'   \n",
       "2               2                               b'1000GENOMES,SSMP,'   \n",
       "3               2                  b'1000GENOMES,WEILL_CORNELL_DGM,'   \n",
       "4               4  b'1000GENOMES,COMPLETE_GENOMICS,SSMP,WEILL_COR...   \n",
       "\n",
       "  alleleFreqCount      alleles                             alleleNs  \\\n",
       "0               3  b'-,A,AAC,'  b'4906.000000,10.000000,92.000000,'   \n",
       "1               2     b'-,CC,'           b'369.000000,4637.000000,'   \n",
       "2               2      b'A,T,'          b'1582.000000,3426.000000,'   \n",
       "3               2      b'C,G,'           b'245.000000,4763.000000,'   \n",
       "4               2      b'A,T,'          b'1585.000000,3423.000000,'   \n",
       "\n",
       "                      alleleFreqs                         bitfields  \n",
       "0  b'0.979633,0.001997,0.018371,'                                {}  \n",
       "1           b'0.073711,0.926288,'                  {maf-5-some-pop}  \n",
       "2           b'0.315895,0.684105,'  {maf-5-some-pop, maf-5-all-pops}  \n",
       "3           b'0.048922,0.951078,'                  {maf-5-some-pop}  \n",
       "4           b'0.316494,0.683506,'  {maf-5-some-pop, maf-5-all-pops}  \n",
       "\n",
       "[5 rows x 26 columns]"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Getting data into pandas:\n",
    "import pandas as pd\n",
    "\n",
    "expr = select([snp_table]).where(snp_table.c.chrom == 'chrY').limit(5)\n",
    "print(expr)\n",
    "pd.read_sql(expr, engine)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false,
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NM_001126115\n",
      "rs17881850 chr17 7669738 7669739 ['C', 'T']\n",
      "rs17880847 chr17 7670578 7670579 ['A', 'T']\n",
      "rs17883852 chr17 7670996 7670997 ['C', 'G']\n",
      "rs17879377 chr17 7671402 7671403 ['A', 'G']\n",
      "rs1641549 chr17 7671456 7671457 ['A', 'G']\n",
      "rs1641548 chr17 7671460 7671461 ['A', 'G']\n",
      "rs858528 chr17 7671617 7671618 ['A', 'G']\n",
      "rs1642793 chr17 7672245 7672246 ['A', 'G']\n",
      "rs1642791 chr17 7673103 7673104 ['C', 'T']\n",
      "rs1800899 chr17 7673522 7673523 ['C', 'T']\n",
      "rs17880604 chr17 7674325 7674326 ['C', 'G']\n",
      "rs1625895 chr17 7674796 7674797 ['A', 'C', 'G']\n",
      "rs1800372 chr17 7674891 7674892 ['A', 'G']\n",
      "NM_001126116\n",
      "rs1800899 chr17 7673522 7673523 ['C', 'T']\n",
      "rs17880604 chr17 7674325 7674326 ['C', 'G']\n",
      "rs1625895 chr17 7674796 7674797 ['A', 'C', 'G']\n",
      "rs1800372 chr17 7674891 7674892 ['A', 'G']\n",
      "NM_001126117\n",
      "rs1800899 chr17 7673522 7673523 ['C', 'T']\n",
      "rs17880604 chr17 7674325 7674326 ['C', 'G']\n",
      "rs1625895 chr17 7674796 7674797 ['A', 'C', 'G']\n",
      "rs1800372 chr17 7674891 7674892 ['A', 'G']\n",
      "NM_001276697\n",
      "rs17881850 chr17 7669738 7669739 ['C', 'T']\n",
      "rs17880847 chr17 7670578 7670579 ['A', 'T']\n",
      "rs17883852 chr17 7670996 7670997 ['C', 'G']\n",
      "rs17879377 chr17 7671402 7671403 ['A', 'G']\n",
      "rs1641549 chr17 7671456 7671457 ['A', 'G']\n",
      "rs1641548 chr17 7671460 7671461 ['A', 'G']\n",
      "rs858528 chr17 7671617 7671618 ['A', 'G']\n",
      "rs1642793 chr17 7672245 7672246 ['A', 'G']\n",
      "rs1642791 chr17 7673103 7673104 ['C', 'T']\n",
      "rs1800899 chr17 7673522 7673523 ['C', 'T']\n",
      "rs17880604 chr17 7674325 7674326 ['C', 'G']\n",
      "rs1625895 chr17 7674796 7674797 ['A', 'C', 'G']\n",
      "rs1800372 chr17 7674891 7674892 ['A', 'G']\n",
      "NM_001276698\n",
      "rs1800899 chr17 7673522 7673523 ['C', 'T']\n",
      "rs17880604 chr17 7674325 7674326 ['C', 'G']\n",
      "rs1625895 chr17 7674796 7674797 ['A', 'C', 'G']\n",
      "rs1800372 chr17 7674891 7674892 ['A', 'G']\n",
      "NM_001276699\n",
      "rs1800899 chr17 7673522 7673523 ['C', 'T']\n",
      "rs17880604 chr17 7674325 7674326 ['C', 'G']\n",
      "rs1625895 chr17 7674796 7674797 ['A', 'C', 'G']\n",
      "rs1800372 chr17 7674891 7674892 ['A', 'G']\n",
      "NM_000546\n",
      "rs17881850 chr17 7669738 7669739 ['C', 'T']\n",
      "rs17880847 chr17 7670578 7670579 ['A', 'T']\n",
      "rs17883852 chr17 7670996 7670997 ['C', 'G']\n",
      "rs17879377 chr17 7671402 7671403 ['A', 'G']\n",
      "rs1641549 chr17 7671456 7671457 ['A', 'G']\n",
      "rs1641548 chr17 7671460 7671461 ['A', 'G']\n",
      "rs858528 chr17 7671617 7671618 ['A', 'G']\n",
      "rs1642793 chr17 7672245 7672246 ['A', 'G']\n",
      "rs1642791 chr17 7673103 7673104 ['C', 'T']\n",
      "rs1800899 chr17 7673522 7673523 ['C', 'T']\n",
      "rs17880604 chr17 7674325 7674326 ['C', 'G']\n",
      "rs1625895 chr17 7674796 7674797 ['A', 'C', 'G']\n",
      "rs1800372 chr17 7674891 7674892 ['A', 'G']\n",
      "rs2909430 chr17 7675326 7675327 ['A', 'G']\n",
      "rs1794287 chr17 7675518 7675519 ['C', 'T']\n",
      "rs1794286 chr17 7675892 7675893 ['A', 'T']\n",
      "rs1042522 chr17 7676153 7676154 ['C', 'G']\n",
      "rs1800370 chr17 7676260 7676261 ['A', 'G']\n",
      "rs17883323 chr17 7676300 7676301 ['A', 'C']\n",
      "rs1642785 chr17 7676482 7676483 ['A', 'C', 'G']\n",
      "NM_001126112\n",
      "rs17881850 chr17 7669738 7669739 ['C', 'T']\n",
      "rs17880847 chr17 7670578 7670579 ['A', 'T']\n",
      "rs17883852 chr17 7670996 7670997 ['C', 'G']\n",
      "rs17879377 chr17 7671402 7671403 ['A', 'G']\n",
      "rs1641549 chr17 7671456 7671457 ['A', 'G']\n",
      "rs1641548 chr17 7671460 7671461 ['A', 'G']\n",
      "rs858528 chr17 7671617 7671618 ['A', 'G']\n",
      "rs1642793 chr17 7672245 7672246 ['A', 'G']\n",
      "rs1642791 chr17 7673103 7673104 ['C', 'T']\n",
      "rs1800899 chr17 7673522 7673523 ['C', 'T']\n",
      "rs17880604 chr17 7674325 7674326 ['C', 'G']\n",
      "rs1625895 chr17 7674796 7674797 ['A', 'C', 'G']\n",
      "rs1800372 chr17 7674891 7674892 ['A', 'G']\n",
      "rs2909430 chr17 7675326 7675327 ['A', 'G']\n",
      "rs1794287 chr17 7675518 7675519 ['C', 'T']\n",
      "rs1794286 chr17 7675892 7675893 ['A', 'T']\n",
      "rs1042522 chr17 7676153 7676154 ['C', 'G']\n",
      "rs1800370 chr17 7676260 7676261 ['A', 'G']\n",
      "rs17883323 chr17 7676300 7676301 ['A', 'C']\n",
      "rs1642785 chr17 7676482 7676483 ['A', 'C', 'G']\n",
      "NM_001126113\n",
      "rs1800899 chr17 7673522 7673523 ['C', 'T']\n",
      "rs17880604 chr17 7674325 7674326 ['C', 'G']\n",
      "rs1625895 chr17 7674796 7674797 ['A', 'C', 'G']\n",
      "rs1800372 chr17 7674891 7674892 ['A', 'G']\n",
      "rs2909430 chr17 7675326 7675327 ['A', 'G']\n",
      "rs1794287 chr17 7675518 7675519 ['C', 'T']\n",
      "rs1794286 chr17 7675892 7675893 ['A', 'T']\n",
      "rs1042522 chr17 7676153 7676154 ['C', 'G']\n",
      "rs1800370 chr17 7676260 7676261 ['A', 'G']\n",
      "rs17883323 chr17 7676300 7676301 ['A', 'C']\n",
      "rs1642785 chr17 7676482 7676483 ['A', 'C', 'G']\n",
      "NM_001126114\n",
      "rs1800899 chr17 7673522 7673523 ['C', 'T']\n",
      "rs17880604 chr17 7674325 7674326 ['C', 'G']\n",
      "rs1625895 chr17 7674796 7674797 ['A', 'C', 'G']\n",
      "rs1800372 chr17 7674891 7674892 ['A', 'G']\n",
      "rs2909430 chr17 7675326 7675327 ['A', 'G']\n",
      "rs1794287 chr17 7675518 7675519 ['C', 'T']\n",
      "rs1794286 chr17 7675892 7675893 ['A', 'T']\n",
      "rs1042522 chr17 7676153 7676154 ['C', 'G']\n",
      "rs1800370 chr17 7676260 7676261 ['A', 'G']\n",
      "rs17883323 chr17 7676300 7676301 ['A', 'C']\n",
      "rs1642785 chr17 7676482 7676483 ['A', 'C', 'G']\n",
      "NM_001126118\n",
      "rs17881850 chr17 7669738 7669739 ['C', 'T']\n",
      "rs17880847 chr17 7670578 7670579 ['A', 'T']\n",
      "rs17883852 chr17 7670996 7670997 ['C', 'G']\n",
      "rs17879377 chr17 7671402 7671403 ['A', 'G']\n",
      "rs1641549 chr17 7671456 7671457 ['A', 'G']\n",
      "rs1641548 chr17 7671460 7671461 ['A', 'G']\n",
      "rs858528 chr17 7671617 7671618 ['A', 'G']\n",
      "rs1642793 chr17 7672245 7672246 ['A', 'G']\n",
      "rs1642791 chr17 7673103 7673104 ['C', 'T']\n",
      "rs1800899 chr17 7673522 7673523 ['C', 'T']\n",
      "rs17880604 chr17 7674325 7674326 ['C', 'G']\n",
      "rs1625895 chr17 7674796 7674797 ['A', 'C', 'G']\n",
      "rs1800372 chr17 7674891 7674892 ['A', 'G']\n",
      "rs2909430 chr17 7675326 7675327 ['A', 'G']\n",
      "rs1794287 chr17 7675518 7675519 ['C', 'T']\n",
      "rs1794286 chr17 7675892 7675893 ['A', 'T']\n",
      "rs1042522 chr17 7676153 7676154 ['C', 'G']\n",
      "NM_001276695\n",
      "rs1800899 chr17 7673522 7673523 ['C', 'T']\n",
      "rs17880604 chr17 7674325 7674326 ['C', 'G']\n",
      "rs1625895 chr17 7674796 7674797 ['A', 'C', 'G']\n",
      "rs1800372 chr17 7674891 7674892 ['A', 'G']\n",
      "rs2909430 chr17 7675326 7675327 ['A', 'G']\n",
      "rs1794287 chr17 7675518 7675519 ['C', 'T']\n",
      "rs1794286 chr17 7675892 7675893 ['A', 'T']\n",
      "rs1042522 chr17 7676153 7676154 ['C', 'G']\n",
      "NM_001276696\n",
      "rs1800899 chr17 7673522 7673523 ['C', 'T']\n",
      "rs17880604 chr17 7674325 7674326 ['C', 'G']\n",
      "rs1625895 chr17 7674796 7674797 ['A', 'C', 'G']\n",
      "rs1800372 chr17 7674891 7674892 ['A', 'G']\n",
      "rs2909430 chr17 7675326 7675327 ['A', 'G']\n",
      "rs1794287 chr17 7675518 7675519 ['C', 'T']\n",
      "rs1794286 chr17 7675892 7675893 ['A', 'T']\n",
      "rs1042522 chr17 7676153 7676154 ['C', 'G']\n",
      "NM_001276760\n",
      "rs17881850 chr17 7669738 7669739 ['C', 'T']\n",
      "rs17880847 chr17 7670578 7670579 ['A', 'T']\n",
      "rs17883852 chr17 7670996 7670997 ['C', 'G']\n",
      "rs17879377 chr17 7671402 7671403 ['A', 'G']\n",
      "rs1641549 chr17 7671456 7671457 ['A', 'G']\n",
      "rs1641548 chr17 7671460 7671461 ['A', 'G']\n",
      "rs858528 chr17 7671617 7671618 ['A', 'G']\n",
      "rs1642793 chr17 7672245 7672246 ['A', 'G']\n",
      "rs1642791 chr17 7673103 7673104 ['C', 'T']\n",
      "rs1800899 chr17 7673522 7673523 ['C', 'T']\n",
      "rs17880604 chr17 7674325 7674326 ['C', 'G']\n",
      "rs1625895 chr17 7674796 7674797 ['A', 'C', 'G']\n",
      "rs1800372 chr17 7674891 7674892 ['A', 'G']\n",
      "rs2909430 chr17 7675326 7675327 ['A', 'G']\n",
      "rs1794287 chr17 7675518 7675519 ['C', 'T']\n",
      "rs1794286 chr17 7675892 7675893 ['A', 'T']\n",
      "rs1042522 chr17 7676153 7676154 ['C', 'G']\n",
      "NM_001276761\n",
      "rs17881850 chr17 7669738 7669739 ['C', 'T']\n",
      "rs17880847 chr17 7670578 7670579 ['A', 'T']\n",
      "rs17883852 chr17 7670996 7670997 ['C', 'G']\n",
      "rs17879377 chr17 7671402 7671403 ['A', 'G']\n",
      "rs1641549 chr17 7671456 7671457 ['A', 'G']\n",
      "rs1641548 chr17 7671460 7671461 ['A', 'G']\n",
      "rs858528 chr17 7671617 7671618 ['A', 'G']\n",
      "rs1642793 chr17 7672245 7672246 ['A', 'G']\n",
      "rs1642791 chr17 7673103 7673104 ['C', 'T']\n",
      "rs1800899 chr17 7673522 7673523 ['C', 'T']\n",
      "rs17880604 chr17 7674325 7674326 ['C', 'G']\n",
      "rs1625895 chr17 7674796 7674797 ['A', 'C', 'G']\n",
      "rs1800372 chr17 7674891 7674892 ['A', 'G']\n",
      "rs2909430 chr17 7675326 7675327 ['A', 'G']\n",
      "rs1794287 chr17 7675518 7675519 ['C', 'T']\n",
      "rs1794286 chr17 7675892 7675893 ['A', 'T']\n",
      "rs1042522 chr17 7676153 7676154 ['C', 'G']\n"
     ]
    }
   ],
   "source": [
    "session.rollback()\n",
    "\n",
    "variants = {}\n",
    "\n",
    "for g in session.query(Gene).filter(Gene.name2 == 'TP53').filter(Gene.cdsEnd > Gene.cdsStart).all():\n",
    "    print(g.name)\n",
    "    snps = session.query(SNP).filter(\n",
    "        SNP.snp_class=='single').filter(\n",
    "        SNP.strand==g.strand).filter(\n",
    "        SNP.chrom==g.chrom).filter(\n",
    "        SNP.chromStart >= g.cdsStart).filter(\n",
    "        SNP.chromEnd <= g.cdsEnd).all()\n",
    "    for s in snps:\n",
    "        alleles = s.alleles.decode('utf-8')[:-1].split(\",\")\n",
    "        print(s.name, s.chrom, s.chromStart, s.chromEnd, alleles)\n",
    "        variants[(s.chrom, s.chromStart)] = alleles\n",
    "    # break # analyze only one gene record, skip the rest - for testing\n",
    "\n",
    "# var = u'jdshjdsjh'\n",
    "# var = b'hshgsdhg'\n",
    "# var.decode('asci')\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import requests, sys, json\n",
    "\n",
    "def get_genome_sequence_ensembl(chrom, start, end):\n",
    "    \"\"\"\n",
    "    API described here http://rest.ensembl.org/documentation/info/sequence_region\n",
    "    \"\"\"\n",
    "    url = 'https://rest.ensembl.org/sequence/region/human/{0}:{1}..{2}:1?content-type=application/json'.format(chrom, start, end)\n",
    "    r = requests.get(url, headers={\"Content-Type\": \"application/json\"}, timeout=10.000)\n",
    "    if not r.ok:\n",
    "        print(\"REST Request FAILED\")\n",
    "        decoded = r.json()\n",
    "        print(decoded['error'])\n",
    "        return\n",
    "    else:\n",
    "        print(\"REST Request OK\")\n",
    "        decoded = r.json()\n",
    "        return decoded['seq']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "# def get_genome_sequence_ncbi(accession, strand):  \n",
    "#     url = \"https://eutils.ncbi.nlm.nih.gov/entrez/eutils/efetch.fcgi?db=nuccore&id={}&strand=1&rettype=fasta&retmode=text\".format(\n",
    "#             accession)     # &seq_start=1&seq_stop=100\n",
    "#     print(url)\n",
    "#     r = requests.get(url)\n",
    "#     if not r.ok:\n",
    "#         print(\"NCBI Request FAILED\")\n",
    "#         print(r.text)\n",
    "#         return\n",
    "#     else:\n",
    "#         print(\"NCBI Request OK\")\n",
    "#         return r.text"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'variants' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-2-656a2641035a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mchrom\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvariants\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0mpositions\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mvariants\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mkeys\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0mstart\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmin\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpositions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0mend\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mmax\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mpositions\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'variants' is not defined"
     ]
    }
   ],
   "source": [
    "chrom = list(variants.keys())[0][0]\n",
    "positions = [k[1] for k in variants.keys()]\n",
    "start = min(positions)\n",
    "end = max(positions)\n",
    "\n",
    "sequence = get_genome_sequence_ensembl(chrom, start, end)\n",
    "print(sequence)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'sequence' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-77f3fc18215e>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0mio\u001b[0m \u001b[0;32mimport\u001b[0m \u001b[0mStringIO\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 6\u001b[0;31m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"Sequence Length\"\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      7\u001b[0m \u001b[0mseqs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m \u001b[0mseqs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m\">\\n\"\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0msequence\u001b[0m\u001b[0;34m+\u001b[0m\u001b[0;34m\"\\n\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'sequence' is not defined"
     ]
    }
   ],
   "source": [
    "from IPython.display import Image\n",
    "\n",
    "from weblogolib import *\n",
    "from io import StringIO\n",
    "\n",
    "print(\"Sequence Length\", len(sequence))\n",
    "seqs = \"\"\n",
    "seqs = \">\\n\"+sequence+\"\\n\"\n",
    "for (chrom, position), alleles in variants.items():\n",
    "    for allele in alleles:\n",
    "        seq = [\"-\"]*len(sequence)\n",
    "        seq[start - position] = allele\n",
    "        seqs += \">\\n\" + \"\".join(seq) + \"\\n\"\n",
    "\n",
    "sequences = read_seq_data(StringIO(seqs))\n",
    "data = LogoData.from_seqs(sequences)\n",
    "options = LogoOptions()\n",
    "options.title = 'A Logo Title'\n",
    "options.scale_width = False\n",
    "options.logo_end = 500\n",
    "options.stacks_per_line = 100\n",
    "formatting = LogoFormat(data, options)\n",
    "\n",
    "png = png_formatter(data, formatting)\n",
    "Image(png)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "anaconda-cloud": {},
  "kernelspec": {
   "display_name": "Python [default]",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.5.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
